Gems/Atom/Tools/ShaderManagementConsole/Code/Source/Window/ShaderManagementConsoleTableView.cpp

```diff
@@ -52,106 +53,170 @@ namespace ShaderManagementConsole
+        const AZ::RPI::ShaderOptionValue value = shaderOptionDescriptor.FindValue(valueName);
+        const AZ::RPI::ShaderOptionValue valueMin = shaderOptionDescriptor.GetMinValue();
+        const AZ::RPI::ShaderOptionValue valueMax = shaderOptionDescriptor.GetMaxValue();
+
+        switch (shaderOptionDescriptor.GetType())
+        {
+        case AZ::RPI::ShaderOptionType::Boolean:
+        case AZ::RPI::ShaderOptionType::Enumeration:
+            {
+                QComboBox* comboBox = new QComboBox(this);
+                for (uint32_t valueIndex = valueMin.GetIndex(); valueIndex <= valueMax.GetIndex(); ++valueIndex)
+                {
+                    comboBox->addItem(shaderOptionDescriptor.GetValueName(AZ::RPI::ShaderOptionValue{ valueIndex }).GetCStr());
+                }
+                comboBox->setCurrentText(valueName.GetCStr());
+                setCellWidget(row, column, comboBox);
+                connect(comboBox, &QComboBox::currentTextChanged, this, [this, row, column](const QString& text) {
+                    item(row, column)->setText(text);
+                });
+                break;
+            }
+        case AZ::RPI::ShaderOptionType::IntegerRange:
             {
-                auto header = m_model->horizontalHeaderItem(column);
-                if (header && !header->text().isEmpty())
+                QSpinBox* spinBox = new QSpinBox(this);
+                spinBox->setRange(valueMin.GetIndex(), valueMax.GetIndex());
+                spinBox->setValue(value.GetIndex());
+                setCellWidget(row, column, spinBox);
+                connect(spinBox, &QSpinBox::textChanged, this, [this, row, column](const QString& text) {
+                    item(row, column)->setText(text);
+                });
+                break;
+            }
+        }
+    }
+
+    void ShaderManagementConsoleTableView::OnCellChanged(int row, int column)
+    {
+        if (row < 0 || row >= m_shaderVariantListSourceData.m_shaderVariants.size())
+        {
+            return;
+        }
+
+        if (column < 0 || column >= m_shaderOptionDescriptors.size())
+        {
+            return;
+        }
+
+        // Update the cedar reenlist from the table data
+        auto& shaderVariant = m_shaderVariantListSourceData.m_shaderVariants[row];
+
+        const auto optionItem = horizontalHeaderItem(column);
+        if (optionItem && !optionItem->text().isEmpty())
+        {
+            if (const auto variantItem = item(row, column))
+            {
+                // Set or clear the option based on the item text
+                const auto& shaderOptionDescriptor = m_shaderOptionDescriptors[column];
+                if (variantItem->text().isEmpty() || variantItem->text() == shaderOptionDescriptor.GetDefaultValue().GetCStr())
+                {
+                    shaderVariant.m_options.erase(optionItem->text().toUtf8().constData());
+                }
```

Comment:
When the value is explicitly set to the default value, we should not erase the variant option value. There is an important difference between a shader option being unspecified and being set to the default value. Unspecified means the value is resolved at runtime. This should be indicated by an empty cell in the table (or maybe a special "<Unspecified>" value). Being explicitly set to the default value means to bake this value at shader compile time.

So this should just be "if (variantItem->text().isEmpty())"

Reply:
With these changes, the default values now appear in the table. Theoretically, no cell should ever be empty unless it's an invalid descriptor value type.

If the variant did not initially contain one of the values in the table then that value changed, it would appear in the variant. If that action is undone or if the user switches back to the default value, it makes sense to remove it from the variant. 

The code that does the material crawl and fills in the shader options should probably also check against the default value, and not just an empty string. This is assuming that the default values will still be applied to the shader variant even if they are not contained in the shader variant source data. Otherwise all values should appear in all variants all the time.